### DISCUSSION TITLE
Feature Request: Reading SAS files

### DISCUSSION THREAD

**Discussion by StaffanBetner** [duckdb/duckdb#10801] [Ideas]
I often work with large data exports from national research registers, where the deliveries from the registry holder typically are in SAS format. There is MIT licence code in the R package haven that can read SAS files, and I guess the code is adaptable to duckdb.

**Comment by rsund** [comment]
I have exactly similar use case. So far I have been able to convert SAS-files to .parquet with DuckDB R-client, but I have investigated how to get the reading possibilities directly to DuckDB. This is obviously something that could be implemented as an extension. I think that haven is a bit slow and it cannot read compressed SAS-files. Better options could be  https://github.com/olivia76/cpp-sas7bdat or https://github.com/jonashaag/sas7bdat.

Unfortunately, I have not had time to work on implementation. One reason has been that the need is most acute in closed remote access environments (similar than what European Health Data Space EHDS will bring to all EU countries) and it would be really difficult to get extensions into such systems unless they would be distributed as R CRAN or Python packages, but it has been really difficult to compile DuckDB extensions with R toolchain and getting it to work so that extensions could be installed using R packages thst would be available in CRAN is even more challenging.

In any case, I strongy support this feature request. It would be optimal, if e.g. @olivia76 or @jonashaag would be interested to provide their excellent sas7bdat reader implementation(s) also as a DuckDB extension!

**Reply by olivia76** [reply]
Hi, sorry for this late answer.  I'm quite busy for the moment.  What kind of collaboration are you looking for?

**Reply by rsund** [reply]
Thank you for answering! If you would happen to have time to take a look would it be feasible (in your opinion) to provide a sink for DuckDB and how to distribute your sas7bdat reader also as a DuckDB extension, that would be awesome. General information about [DuckDB extensions here](https://duckdb.org/docs/extensions/overview.html) and an [extension-template repository here](https://github.com/duckdb/extension-template/). If you think that it would require only a feasible amount of (hopefully straightforward) work, then only the actual implementation would be missing ;). I could try to help, but I'm not an expert in C++ and my time is also limited. And I guess that all this would be voluntary work for the benefit of the community.

**Reply by olivia76** [reply]
Hi, I will look at it after coming back from vacations end of July.  Your default c++ package manager is vcpkg. for my project I used conan2.  I guess it would be possible to migrate from one to the other.

**Comment by scarnecchia** [comment]
I agree that this would be quite useful and similarly to @rsund, find the need is most acute within public health surveillance / pharmacoepidemiology spaces where dependence on SAS is driven by legacy and compliance reasons but the datasets are frequently too large for haven to handle.

**Comment by emraher** [comment]
https://github.com/duckdb/duckdb/discussions/14504

**Comment by dylanmeysmans** [comment]
> I often work with large data exports from national research registers, where the deliveries from the registry holder typically are in SAS format. There is MIT licence code in the R package haven that can read SAS files, and I guess the code is adaptable to duckdb.

@emraher  @rsund @olivia76 @StaffanBetner @scarnecchia I managed to implement this as a community extension, see https://github.com/mettekou/duckdb-read-stat.

**Reply by StaffanBetner** [reply]
@mettekou Wow! Will try it out!
